From 112fe87956d75dbe83483692a523881447777ba4 Mon Sep 17 00:00:00 2001
From: Wenbinc-Bin <wenbin.chen@intel.com>
Date: Mon, 18 Jan 2021 17:08:39 +0800
Subject: [PATCH 60/88] libavcodec/qsvdec: move unref before get_format

Qsv decoder needs to call preinit() twice to get NumFrameSuggest
to config init_pool_size. The hw_frame_ctx is released at the
second time calling preinit(). The hw_frame_ctx is AVBufferRef, so
it release the source when the last unref is called. When the old
context is released, the new one is already created, and somehow the device
memory is stil be taken, even if the destory memory function is called.
Moving unref before creating new context can solve this problem.

Signed-off-by Wenbin Chen <wenbin.chen@intel.com>
---
 libavcodec/qsvdec.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libavcodec/qsvdec.c b/libavcodec/qsvdec.c
index 1d1a208ef3..4c74de68b1 100644
--- a/libavcodec/qsvdec.c
+++ b/libavcodec/qsvdec.c
@@ -236,6 +236,8 @@ static int qsv_decode_preinit(AVCodecContext *avctx, QSVContext *q, enum AVPixel
         pix_fmt,        /* system memory format obtained from bitstream parser */
         AV_PIX_FMT_NONE };
 
+    av_buffer_unref(&q->frames_ctx.mids_buf);
+    av_buffer_unref(&q->frames_ctx.hw_frames_ctx);
     ret = ff_get_format(avctx, pix_fmts);
     if (ret < 0) {
         q->orig_pix_fmt = avctx->pix_fmt = AV_PIX_FMT_NONE;
-- 
2.17.1

